---
title: "Traits that co-vary"
author: "SK"
date: "5/23/2019"
output: html_document
---





```{r setup, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = FALSE, comment = FALSE, 
                      warning = FALSE, message = FALSE)
```


# Simulate data:
  * K = 20 -> number of strains
  * Nk = 5 -> number of individuals per strain
  * N = 20*5 -> number of individuals
  * S = 100 -> number of SNPs
  * Three traits: $Y_1$, $Y_2$, $Y_3$
      * $\alpha = 0$
      * $\pi = Uniform(0.5)$
      * for given strain: if $\pi >= 0.5, \beta = N([0-2], 1)$, else $N([3-5], 1)$
      * $Y_1 = \alpha + \beta$
      * $Y_2 = \alpha + \beta \cdot N(0, 1)$
      * $Y_3 = \alpha + \beta \cdot N(0.5, 0.5)$
  * Generate X (genotypes) with $\pi_{signal} = 0.9$ -> proportion of background SNPs
      * if $\pi_{signal}$ => generate alleles of SNP with probability from $Y$
      * if $\pi_{noise}$ => generate alleles of SNP with probability = 0.5
      



```{r, results = "hide", include = FALSE}
require(rstan)
require(loo)
require(ggplot2)

source("R/main.R")
source("R/util.R")
source("R/bayesian.R")
source("R/statlearn.R")
source("R/modelcomparison.R")
source("R/ppc.R")

# models
m1 <- rstan::stan_model(file = "src/stan_files/M1.stan")
m1c <- rstan::stan_model(file = "src/stan_files/M1c.stan")
```










```{r}
# 1. simulate data
K <- 20
Nk <- 5
N <- K*Nk
S <- 100
p.signal = 0.9


set.seed(seed = 42133214)

# simulate slopes
beta <- replicate(n = K, ifelse(test = rep(x = runif(n = 1, min = 0, max = 1), times = Nk) > 0.5,
                                yes = rnorm(n = Nk, mean = runif(n = 1, min = 0, max = 2), sd = 1),
                                no = rnorm(n = Nk, mean = runif(n = 1, min = 3, max = 5), sd = 1)))

beta.mu <- colMeans(beta)
beta <- as.vector(beta)


alpha <- 0
Y <- alpha + beta



# no trait covariance
cor <- rnorm(n = N, mean = 0, sd = 1)
Y2 <- alpha + beta * cor


# with trait covariance
cor <- rnorm(n = N, mean = 0.5, sd = 0.5)
Y3 <- alpha + beta * cor


# simulate genotypes
X <- matrix(data = NA, nrow = N, ncol = S)
signal.snps <- c()
for(i in 1:S) {
  if(runif(n = 1, min = 0, max = 1) >= p.signal) {
    signal.snps <- c(signal.snps, i)
    X.vec <- sapply(X = (beta.mu-min(beta.mu))/(max(beta.mu)-min(beta.mu)), FUN = rbinom, n = Nk, size = 1)
    X.vec <- as.vector(X.vec)
    if(runif(n = 1, min = 0, max = 1) > 0.5) {
      X.vec[X.vec == 0] <- -1
      X[, i] <- X.vec
    }
    else {
      X.vec[X.vec == 1] <- -1
      X.vec[X.vec == 0] <- 1
      X[, i] <- X.vec
    }
  }
  else {
    X.vec <- replicate(n = K, sapply(X = rep(x = sample(x = c(0, 1), size = 1, replace = F), times = Nk), 
                                     FUN = rbinom, n = 1, size = 1))
    X.vec <- as.vector(X.vec)
    if(runif(n = 1, min = 0, max = 1) > 0.5) {
      X.vec[X.vec == 0] <- -1
      X[, i] <- X.vec
    }
    else {
      X.vec[X.vec == 1] <- -1
      X.vec[X.vec == 0] <- 1
      X[, i] <- X.vec
    }
    X[, i] <- X.vec
  }
}
rm(i)

# pack data for MLgwas
Yq <- matrix(data = cbind(Y, Y2, Y3), ncol = 3)
Yd <- matrix(data = 0, ncol = 0, nrow = N)
strains <- rep(x = 1:K, each = Nk)
data.list <- list(N = N,
                  Nk = K,
                  K = strains,
                  Ntq = ncol(Yq),
                  Ntd = 0,
                  Ns = S,
                  Yq = Yq,
                  Yd = Yd,
                  X = X,
                  p.signal = p.signal,
                  signal.snps = signal.snps,
                  strains = strains)

gt.data <- getStanData(genotype = X,
                       traits = Yq,
                       trait.type = c("Q", "Q"),
                       strains = strains)


d1 <- list(data.list = data.list, 
           gt.data = gt.data)

rm(X, N, S, Y, Yq, Yd, data.list, gt.data, K, 
   strains, Nk, beta.mu, signal.snps, p.signal, 
   beta, alpha, i, X.vec)
```





## Analyze data
  * Comparison old vs new

```{r}
p <- rstan::sampling(object = m1,
                     data = d1$data.list,
                     iter = 1000,
                     warmup = 500,
                     chains = 4,
                     cores = 4,
                     control = list(adapt_delta = 0.9),
                     seed = 43397180)

pc <- rstan::sampling(object = m1c,
                     data = d1$data.list,
                     iter = 1000,
                     warmup = 500,
                     chains = 4,
                     cores = 4,
                     control = list(adapt_delta = 0.9),
                     seed = 43397180)
```



```{r, fig.width=5, fig.height=5}
p.summary <- summary(p, par = "beta")$summary[, "mean"]
pc.summary <- summary(pc, par = "beta")$summary[, "mean"]
plot(p.summary, pc.summary)
abline(0, 1, col = "red")
```





```{r, fig.width=8, fig.height=6}
p.summary <- summary(p, par = "mu_beta")$summary
pc.summary <- summary(pc, par = "mu_beta")$summary

par(mfrow = c(2, 3))
plot(pc.summary[1:100, "mean"], 
     pc.summary[101:200, "mean"],
     xlab = "Y1", ylab = "Y2", main = "Cov+")
abline(0, 1, col = "red")
plot(pc.summary[1:100, "mean"], 
     pc.summary[201:300, "mean"],
     xlab = "Y1", ylab = "Y3", main = "Cov+")
abline(0, 1, col = "red")
plot(pc.summary[101:200, "mean"], 
     pc.summary[201:300, "mean"],
     xlab = "Y2", ylab = "Y3", 
     main = "Cov+")
abline(0, 1, col = "red")

plot(p.summary[1:100, "mean"], 
     p.summary[101:200, "mean"],
     xlab = "Y2", ylab = "Y3", 
     main = "Cov-")
abline(0, 1, col = "red")
plot(p.summary[1:100, "mean"], 
     p.summary[201:300, "mean"],
     xlab = "Y2", ylab = "Y3", 
     main = "Cov-")
abline(0, 1, col = "red")
plot(p.summary[101:200, "mean"], 
     p.summary[201:300, "mean"],
     xlab = "Y2", ylab = "Y3", 
     main = "Cov-")
abline(0, 1, col = "red")
```




```{r, fig.width=8, fig.height=3}
p.summary <- summary(p, par = "mu_beta")$summary
pc.summary <- summary(pc, par = "mu_beta")$summary
plot(p.summary[, 7]-p.summary[, 4], pc.summary[, 7]-pc.summary[, 4])
abline(0, 1, col = "red")

par(mfrow = c(1, 3))
plot(pc.summary[1:100, 7]-pc.summary[1:100, 4], 
     p.summary[1:100, 7]-p.summary[1:100, 4],
     xlab = "Cov+ HDI-length", 
     ylab = "Cov- HDI-length", 
     main = "Y1")
abline(0, 1, col = "red")
plot(pc.summary[101:200, 7]-pc.summary[101:200, 4], 
     p.summary[101:200, 7]-p.summary[101:200, 4],
     xlab = "Cov+ HDI-length", 
     ylab = "Cov- HDI-length", 
     main = "Y2")
abline(0, 1, col = "red")
plot(pc.summary[201:300, 7]-pc.summary[201:300, 4], 
     p.summary[201:300, 7]-p.summary[201:300, 4],
     xlab = "Cov+ HDI-length", 
     ylab = "Cov- HDI-length", 
     main = "Y3")
abline(0, 1, col = "red")
```



```{r}
summary(pc, par = "rho")$summary
```
